File: Problems/Problem_1195_Fizz_Buzz_Multithreaded/go/FizzBuzz.go
package main

import (
	"fmt"      // Paquete para imprimir mensajes en consola
	"sync"     // Paquete para sincronización de goroutines (Mutex, Cond, WaitGroup)
)

// Estructura FizzBuzz
type FizzBuzz struct {
	n       int        // Número límite hasta donde queremos ejecutar FizzBuzz
	counter int        // Contador actual que se va incrementando
	mu      sync.Mutex // Mutex para manejar la exclusión mutua entre goroutines
	cond    *sync.Cond // Condición para manejar señales entre goroutines
}

// Constructor para inicializar una nueva instancia de FizzBuzz
func NewFizzBuzz(n int) *FizzBuzz {
	// Creamos una instancia de FizzBuzz con el límite `n` y el contador inicializado en 1
	fb := &FizzBuzz{n: n, counter: 1}

	// Inicializamos la condición asociada al Mutex
	fb.cond = sync.NewCond(&fb.mu) // `&` obtiene la dirección de memoria del Mutex `mu`
	return fb
}

// Método para imprimir "fizz" si el número actual es divisible por 3 pero no por 5
func (fb *FizzBuzz) fizz(printFizz func()) {
	for { // Bucle infinito que dependerá de condiciones internas para detenerse
		fb.mu.Lock() // Bloqueamos el mutex para garantizar exclusión mutua
		// Mientras el número actual no cumple las condiciones de "fizz"
		for fb.counter <= fb.n && (fb.counter%3 != 0 || fb.counter%5 == 0) {
			fb.cond.Wait() // Esperamos una señal de otras goroutines
		}
		// Si ya hemos procesado todos los números, desbloqueamos y salimos
		if fb.counter > fb.n {
			fb.mu.Unlock()
			return
		}
		// Ejecutamos la función para imprimir "fizz"
		printFizz()
		fb.counter++        // Incrementamos el contador
		fb.cond.Broadcast() // Despertamos a todas las goroutines que están esperando
		fb.mu.Unlock()      // Desbloqueamos el mutex
	}
}

// Método para imprimir "buzz" si el número actual es divisible por 5 pero no por 3
func (fb *FizzBuzz) buzz(printBuzz func()) {
	for {
		fb.mu.Lock()
		// Mientras el número actual no cumple las condiciones de "buzz"
		for fb.counter <= fb.n && (fb.counter%5 != 0 || fb.counter%3 == 0) {
			fb.cond.Wait() // Esperamos una señal de otras goroutines
		}
		if fb.counter > fb.n {
			fb.mu.Unlock()
			return
		}
		printBuzz()
		fb.counter++
		fb.cond.Broadcast() // Notificamos a otras goroutines
		fb.mu.Unlock()
	}
}

// Método para imprimir "fizzbuzz" si el número actual es divisible tanto por 3 como por 5
func (fb *FizzBuzz) fizzbuzz(printFizzBuzz func()) {
	for {
		fb.mu.Lock()
		// Mientras el número actual no cumple las condiciones de "fizzbuzz"
		for fb.counter <= fb.n && (fb.counter%3 != 0 || fb.counter%5 != 0) {
			fb.cond.Wait() // Esperamos una señal
		}
		if fb.counter > fb.n {
			fb.mu.Unlock()
			return
		}
		printFizzBuzz()
		fb.counter++
		fb.cond.Broadcast() // Notificamos a otras goroutines
		fb.mu.Unlock()
	}
}

// Método para imprimir números si no son divisibles por 3 ni por 5
func (fb *FizzBuzz) number(printNumber func(int)) {
	for {
		fb.mu.Lock()
		// Mientras el número actual es divisible por 3 o por 5
		for fb.counter <= fb.n && (fb.counter%3 == 0 || fb.counter%5 == 0) {
			fb.cond.Wait() // Esperamos una señal
		}
		if fb.counter > fb.n {
			fb.mu.Unlock()
			return
		}
		printNumber(fb.counter) // Imprimimos el número actual
		fb.counter++
		fb.cond.Broadcast() // Notificamos a otras goroutines
		fb.mu.Unlock()
	}
}

func main() {
	n := 15                  // Límite hasta el que queremos ejecutar FizzBuzz
	fb := NewFizzBuzz(n)     // Creamos una nueva instancia de FizzBuzz

	var wg sync.WaitGroup    // WaitGroup para esperar a que todas las goroutines terminen
	wg.Add(4)                // Añadimos 4 al contador, una por cada goroutine

	// Lanzamos la goroutine para manejar "fizz"
	go func() {
		defer wg.Done() // Disminuimos el contador del WaitGroup al finalizar
		fb.fizz(func() { fmt.Println("fizz") })
	}()
	// Lanzamos la goroutine para manejar "buzz"
	go func() {
		defer wg.Done()
		fb.buzz(func() { fmt.Println("buzz") })
	}()
	// Lanzamos la goroutine para manejar "fizzbuzz"
	go func() {
		defer wg.Done()
		fb.fizzbuzz(func() { fmt.Println("fizzbuzz") })
	}()
	// Lanzamos la goroutine para manejar números
	go func() {
		defer wg.Done()
		fb.number(func(x int) { fmt.Println(x) })
	}()

	wg.Wait() // Esperamos a que todas las goroutines terminen
}

File: Problems/Problem_1195_Fizz_Buzz_Multithreaded/rust/FizzBuzz.rs
use std::sync::{Arc, Mutex}; // Arc y Mutex para sincronización entre hilos
use std::thread;

// Estructura principal para FizzBuzz
struct FizzBuzz {
    n: usize, // Número máximo hasta el cual se ejecutará FizzBuzz
}

impl FizzBuzz {
    // Constructor para inicializar la estructura
    fn new(n: usize) -> Self {
        Self { n }
    }

    /// Método genérico para manejar una acción basada en una condición
    ///
    /// - `condition`: Un cierre (closure) que toma un usize y retorna un bool.
    /// - `action`: Un cierre (closure) que ejecuta una acción basada en la condición.
    fn run(
        &self,
        condition: impl Fn(usize) -> bool + Send + Sync + 'static,
        action: Arc<dyn Fn() + Send + Sync>, // Usamos Arc para permitir compartir entre hilos
    ) {
        let n = self.n; // Capturamos el valor máximo para el bucle
        thread::spawn(move || {
            for i in 1..=n {
                if condition(i) {
                    action(); // Ejecutamos la acción si la condición se cumple
                }
            }
        })
        .join()
        .expect("El hilo no pudo completarse correctamente.");
    }

    /// Lógica para "Fizz" (números divisibles por 3 pero no por 5)
    fn fizz(&self, print_fizz: Arc<dyn Fn() + Send + Sync>) {
        self.run(
            |x| x % 3 == 0 && x % 5 != 0, // Condición: divisible por 3 pero no por 5
            print_fizz,
        );
    }

    /// Lógica para "Buzz" (números divisibles por 5 pero no por 3)
    fn buzz(&self, print_buzz: Arc<dyn Fn() + Send + Sync>) {
        self.run(
            |x| x % 5 == 0 && x % 3 != 0, // Condición: divisible por 5 pero no por 3
            print_buzz,
        );
    }

    /// Lógica para "FizzBuzz" (números divisibles por 3 y 5)
    fn fizzbuzz(&self, print_fizzbuzz: Arc<dyn Fn() + Send + Sync>) {
        self.run(
            |x| x % 3 == 0 && x % 5 == 0, // Condición: divisible por 3 y 5
            print_fizzbuzz,
        );
    }

    /// Lógica para números que no son divisibles ni por 3 ni por 5
    fn number(&self, print_number: Arc<dyn Fn(usize) + Send + Sync>) {
        let n = self.n;
        thread::spawn(move || {
            for i in 1..=n {
                if i % 3 != 0 && i % 5 != 0 {
                    print_number(i); // Ejecutamos la función para imprimir el número
                }
            }
        })
        .join()
        .expect("El hilo no pudo completarse correctamente.");
    }
}

fn main() {
    // Número máximo hasta el cual se ejecutará FizzBuzz
    let n = 15;

    // Crear una nueva instancia de FizzBuzz
    let fizz_buzz = FizzBuzz::new(n);

    // Definir las acciones (usando closures envueltos en Arc)
    let print_fizz = Arc::new(|| println!("Fizz"));
    let print_buzz = Arc::new(|| println!("Buzz"));
    let print_fizzbuzz = Arc::new(|| println!("FizzBuzz"));
    let print_number = Arc::new(|x: usize| println!("{}", x));

    // Ejecutar cada lógica en threads separados
    fizz_buzz.fizz(print_fizz.clone()); // Lógica para "Fizz"
    fizz_buzz.buzz(print_buzz.clone()); // Lógica para "Buzz"
    fizz_buzz.fizzbuzz(print_fizzbuzz.clone()); // Lógica para "FizzBuzz"
    fizz_buzz.number(print_number.clone()); // Lógica para números restantes
}
